### PR TITLE
【paddle.fleet】add gradient Merge Optimizer

### PR DESCRIPTION
**Pull request by mapingshuo**
### PR types

New features

### PR changes

APIs

### Describe

在分布式训练中，经常遇到显存或者内存不足的情况，这通常有三种原因：

- 中间层输出占据的显存超出了内存/显存大小
- 参数过大（比如CPU的embedding）
- 输入Var过大（比如视频输入）

GradientMerge(梯度累加) 策略的做法，是将大Batch的输入改成小Batch，分别进行“前向+反向”网络计算梯度；最后将梯度做累加。

In the distributed training, we often encounter insufficient memory problem. There are usually 3 reasons:

- The memory occupied by the middle layer output exceeds the memory size
- The parameter is too large (such as CPU embedding)
- Input Var is too large (such as input)

The GradientMerge (gradient accumulation) strategy is to split the input of a large batch to a small batch, it performs a "forward + backward" network to calculate the gradient for each small batch; finally, the gradient is accumulated and the parameter is updated by specific optimize algorithm

### CODE REVIEW

File: python/paddle/fluid/optimizer.py [hunk 1]
```diff
@@ -4929,3 +4929,204 @@ def minimize(self, loss, startup_program=None):
             with switch.default():
                 pass
         return mini_out
+
+
+class GradientMergeOptimizer(object):
+    """
+    Gradient Merge, also called as Gradient Accumulation,
+    is a training strategy for bigger batches. With this strategy,
```
Comment (guru4elephant) [inline python/paddle/fluid/optimizer.py]: larger batch

File: python/paddle/fluid/optimizer.py [hunk 2]
```diff
@@ -4929,3 +4929,204 @@ def minimize(self, loss, startup_program=None):
...
+            prediction = fluid.layers.fc(input=[fc_1], size=label_dim, act='softmax')
+            cost = fluid.layers.cross_entropy(input=prediction, label=input_y)
+            sum_cost = fluid.layers.reduce_mean(cost)
+            return sum_cost, fc_1, prediction
+
+        input_x = fluid.layers.data(name="x", shape=[32], dtype='float32')
+        input_y = fluid.layers.data(name="y", shape=[1], dtype='int64')
+        cost, fc_1, pred = mlp(input_x, input_y)
+        sgd = fluid.optimizer.Adam(learning_rate=0.01)
+        sgd = fluid.optimizer.GradientMergeOptimizer(sgd, k_steps=4, avg=True)
```
Comment (guru4elephant) [inline python/paddle/fluid/optimizer.py:4975]: can you set the arguments with extra functions?  I think current initialization method is not good to generalization for other optimizers.

File: python/paddle/fluid/optimizer.py [hunk 3]
```diff
@@ -4929,3 +4929,204 @@ def minimize(self, loss, startup_program=None):
...
+
+            mod = layers.elementwise_mod(gradient_merge_step, gradient_merge_k)
+            with layers.control_flow.Switch() as switch:
+                with switch.case(mod != zero_var):
+                    # 1. update the gradient_merge_vars
+                    #  gradient_merge_vars += gradient_vars
+                    for param_name in param_names:
+                        grad = param_to_grad[param_name]
+                        grad_merge = param_to_gradient_merge[param_name]
+                        tmp_var = layers.elementwise_add(grad, grad_merge)
```
Comment (guru4elephant) [inline python/paddle/fluid/optimizer.py]: why not inplace?

**Review by guru4elephant — COMMENTED** [review]
please add this to paddle.fleet in your next PRs

File: python/paddle/fluid/optimizer.py (hunk 3 above)
Comment (mapingshuo) [reply python/paddle/fluid/optimizer.py]: Yes, inplace is more efficient, I will optimize it soon.

File: python/paddle/fluid/optimizer.py (hunk 1 above)
Comment (mapingshuo) [reply python/paddle/fluid/optimizer.py]: corrected.

**Comment by mapingshuo** [comment]
> please add this to paddle.fleet in your next PRs

Thanks for your advice, I will add it to paddle.fleet soon.

**Review by guru4elephant — APPROVED** [review]
LGTM
